File: _inc/laravel/app/Http/Controllers/Shapes/JournalEntryController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
    DatabaseConstants,
    MiddlewaresConstants,
    PermissionsConstants,
    ViewsConstants
};
use App\Models\{
    BankAccount,
    ChartOfAccount,
    JournalEntry,
    JournalItem,
    Utility
};
use App\Traits\{ChecksLogin, ChecksPermissions};
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Http\{Request, RedirectResponse, JsonResponse};
use Illuminate\Support\Facades\{Auth, DB, Log, Validator, View as ViewFacade};
use Illuminate\View\View;

class JournalEntryController extends Controller
{
    use ChecksLogin, ChecksPermissions;

    private const INDEX_ROUTE = ViewsConstants::JRN_ET . '.index';

    public function __construct()
    {
        $this->middleware([MiddlewaresConstants::AUTH, MiddlewaresConstants::XSS]);
    }

    public function index(Request $request): View|RedirectResponse|JsonResponse
    {
        $action   = __FUNCTION__;
        $class    = static::class;
        $sig      = "$class::$action";
        $viewPath = ViewsConstants::JRN_ET . '.index';

        return $this->measureProfile($action, function () use ($request, $class, $action, $sig, $viewPath) {
            Log::info("$sig start", ['user' => Auth::id()]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, 'manage journal entry', self::INDEX_ROUTE)) !== true) {
                Log::warning("$sig denied", ['user' => Auth::id()]);
                return $c;
            }

            try {
                $t = microtime(true);
                $entries = JournalEntry::where(DatabaseConstants::TABLE_CREATOR, $user?->creatorId())->get();
                $this->logExecutionTime($t, $sig, 'fetchEntries');
                Log::info("$sig fetched", ['count' => $entries->count()]);

                $t = microtime(true);
                if (!ViewFacade::exists($viewPath)) {
                    $this->logExecutionTime($t, $sig, 'viewExistsCheck');
                    return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                }
                $this->logExecutionTime($t, $sig, 'viewExistsCheck');

                return view($viewPath, compact('entries'));
            } catch (\Throwable $e) {
                Log::error("$sig error", ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, "$class::$action");
            }
        });
    }

    public function create(Request $request): View|JsonResponse|RedirectResponse
    {
        $action   = __FUNCTION__;
        $class    = static::class;
        $sig      = "$class::$action";
        $viewPath = ViewsConstants::JRN_ET . '.create';

        return $this->measureProfile($action, function () use ($request, $class, $action, $sig, $viewPath) {
            Log::info("$sig start", ['user' => Auth::id()]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, 'create journal entry', self::INDEX_ROUTE)) !== true) {
                Log::warning("$sig denied", ['user' => Auth::id()]);
                return $c;
            }

            $t = microtime(true);
            $accounts = ChartOfAccount::selectRaw("CONCAT(code,' - ',name) AS code_name, id")
                ->where(DatabaseConstants::TABLE_CREATOR, $user?->creatorId())
                ->pluck('code_name', 'id');
            $this->logExecutionTime($t, $sig, 'pluckChartOfAccounts');

            $t = microtime(true);
            $journalId = $this->journalNumber(); // login already checked
            $this->logExecutionTime($t, $sig, 'generateJournalNumber');

            $t = microtime(true);
            if (!ViewFacade::exists($viewPath)) {
                $this->logExecutionTime($t, $sig, 'viewExistsCheck');
                return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
            }
            $this->logExecutionTime($t, $sig, 'viewExistsCheck');

            return view($viewPath, compact('accounts', 'journalId'));
        });
    }

    public function store(Request $request): RedirectResponse
    {
        $action = __FUNCTION__;
        $class  = static::class;
        $sig    = "$class::$action";

        return $this->measureProfile($action, function () use ($request, $class, $action, $sig) {
            Log::info("$sig start", ['input' => $request->only('date', 'accounts')]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, 'create journal entry', self::INDEX_ROUTE)) !== true) {
                Log::warning("$sig denied", ['user' => Auth::id()]);
                return $c;
            }

            $t = microtime(true);
            $validator = Validator::make($request->all(), [
                'date'     => 'required|date',
                'accounts' => 'required|array|min:1'
            ]);
            if ($validator->fails()) {
                $this->logExecutionTime($t, $sig, 'validateFail');
                Log::warning("$sig validation failed", ['err' => $validator->errors()->first()]);
                return redirect()->back()->with('error', $validator->errors()->first());
            }
            $this->logExecutionTime($t, $sig, 'validateSuccess');

            $t = microtime(true);
            $accounts = $request->input('accounts');
            $totals = array_reduce($accounts, function ($tot, $item) {
                $tot['debit']  += $item['debit']  ?? 0;
                $tot['credit'] += $item['credit'] ?? 0;
                return $tot;
            }, ['debit' => 0, 'credit' => 0]);
            $this->logExecutionTime($t, $sig, 'computeTotals');

            if ($totals['debit'] !== $totals['credit']) {
                Log::warning("$sig imbalanced entry", $totals);
                return redirect()->back()->with('error', __('Debit and Credit must be Equal.'));
            }

            try {
                $t = microtime(true);
                DB::transaction(function () use ($request, $accounts, $user, $sig) {
                    $journal = JournalEntry::create([
                        'journal_id'  => $this->journalNumber(),
                        'date'        => $request->date,
                        'reference'   => $request->reference,
                        'description' => $request->description,
                        DatabaseConstants::TABLE_CREATOR => $user?->creatorId(),
                    ]);
                    Log::info("$sig created journal", ['id' => $journal->id]);

                    foreach ($accounts as $item) {
                        $journalItem = $journal->items()->create([
                            'account'     => $item['account'],
                            'description' => $item['description'] ?? null,
                            'debit'       => $item['debit']  ?? 0,
                            'credit'      => $item['credit'] ?? 0,
                        ]);
                        Log::info("$sig created item", ['item_id' => $journalItem->id]);
                        $this->updateBankBalances($journalItem);
                    }
                });
                $this->logExecutionTime($t, $sig, 'transactionCommit');

                return redirect()->route(self::INDEX_ROUTE)
                    ->with('success', __('Journal entry successfully created.'));
            } catch (\Throwable $e) {
                Log::error("$sig error", ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, "$class::$action");
            }
        });
    }

    public function show(JournalEntry $journalEntry): View|RedirectResponse
    {
        $action   = __FUNCTION__;
        $class    = static::class;
        $sig      = "$class::$action";
        $viewPath = ViewsConstants::JRN_ET . '.view';

        return $this->measureProfile($action, function () use ($journalEntry, $class, $action, $sig, $viewPath) {
            Log::info("$sig start", ['id' => $journalEntry->id]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if ($c = self::guard(request(), 'show journal entry', self::INDEX_ROUTE)) {
                Log::warning("$sig denied", ['user' => Auth::id()]);
                return $c;
            }
            if ($journalEntry[DatabaseConstants::TABLE_CREATOR] !== $user?->creatorId()) {
                Log::warning("$sig forbidden", ['owner' => $journalEntry[DatabaseConstants::TABLE_CREATOR]]);
                return defaultPermissionDenial(request(), new AuthorizationException(), "$class::$action");
            }

            $t = microtime(true);
            $accounts = $journalEntry->items;
            $settings = Utility::settings();
            $this->logExecutionTime($t, $sig, 'loadRelationsAndSettings');

            $t = microtime(true);
            if (!ViewFacade::exists($viewPath)) {
                $this->logExecutionTime($t, $sig, 'viewExistsCheck');
                return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
            }
            $this->logExecutionTime($t, $sig, 'viewExistsCheck');

            return view($viewPath, compact('journalEntry', 'accounts', 'settings'));
        });
    }

    public function edit(JournalEntry $journalEntry): View|JsonResponse|RedirectResponse
    {
        $action   = __FUNCTION__;
        $class    = static::class;
        $sig      = "$class::$action";
        $viewPath = ViewsConstants::JRN_ET . '.edit';

        return $this->measureProfile($action, function () use ($journalEntry, $class, $action, $sig, $viewPath) {
            Log::info("$sig start", ['id' => $journalEntry->id]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if ($c = self::guard(request(), 'edit journal entry', self::INDEX_ROUTE)) {
                Log::warning("$sig denied", ['user' => Auth::id()]);
                return $c;
            }
            if ($journalEntry[DatabaseConstants::TABLE_CREATOR] !== $user?->creatorId()) {
                Log::warning("$sig forbidden", ['owner' => $journalEntry[DatabaseConstants::TABLE_CREATOR]]);
                return defaultPermissionDenial(request(), new AuthorizationException(), "$class::$action");
            }

            $t = microtime(true);
            $accounts = ChartOfAccount::selectRaw("CONCAT(code,' - ',name) AS code_name, id")
                ->where(DatabaseConstants::TABLE_CREATOR, $user?->creatorId())
                ->pluck('code_name', 'id');
            $this->logExecutionTime($t, $sig, 'pluckChartOfAccounts');

            $t = microtime(true);
            if (!ViewFacade::exists($viewPath)) {
                $this->logExecutionTime($t, $sig, 'viewExistsCheck');
                return redirect()->back()->with('error', "HTTP 404: Page {$viewPath} not found!");
            }
            $this->logExecutionTime($t, $sig, 'viewExistsCheck');

            return view($viewPath, compact('accounts', 'journalEntry'));
        });
    }

    public function update(Request $request, JournalEntry $journalEntry): RedirectResponse
    {
        $action = __FUNCTION__;
        $class  = static::class;
        $sig    = "$class::$action";

        return $this->measureProfile($action, function () use ($request, $journalEntry, $class, $action, $sig) {
            Log::info("$sig start", ['id' => $journalEntry->id, 'input' => $request->only('date', 'accounts')]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, 'edit journal entry', self::INDEX_ROUTE)) !== true) {
                Log::warning("$sig denied", ['user' => Auth::id()]);
                return $c;
            }
            if ($journalEntry[DatabaseConstants::TABLE_CREATOR] !== $user?->creatorId()) {
                Log::warning("$sig forbidden", ['owner' => $journalEntry[DatabaseConstants::TABLE_CREATOR]]);
                return defaultPermissionDenial($request, new AuthorizationException(), "$class::$action");
            }

            $t = microtime(true);
            $validator = Validator::make($request->all(), [
                'date'     => 'required|date',
                'accounts' => 'required|array|min:1'
            ]);
            if ($validator->fails()) {
                $this->logExecutionTime($t, $sig, 'validateFail');
                Log::warning("$sig validation failed", ['err' => $validator->errors()->first()]);
                return redirect()->back()->with('error', $validator->errors()->first());
            }
            $this->logExecutionTime($t, $sig, 'validateSuccess');

            $t = microtime(true);
            $accounts = $request->input('accounts');
            $totals = array_reduce($accounts, function ($tot, $item) {
                $tot['debit']  += $item['debit']  ?? 0;
                $tot['credit'] += $item['credit'] ?? 0;
                return $tot;
            }, ['debit' => 0, 'credit' => 0]);
            $this->logExecutionTime($t, $sig, 'computeTotals');

            if ($totals['debit'] !== $totals['credit']) {
                Log::warning("$sig imbalanced", $totals);
                return redirect()->back()->with('error', __('Debit and Credit must be Equal.'));
            }

            try {
                $t = microtime(true);
                DB::transaction(function () use ($request, $journalEntry, $accounts, $sig) {
                    $journalEntry->update([
                        'date'        => $request->date,
                        'reference'   => $request->reference,
                        'description' => $request->description,
                    ]);
                    Log::info("$sig journal updated", ['id' => $journalEntry->id]);

                    $existingIds = collect($accounts)->pluck('id')->filter()->all();
                    JournalItem::where('journal', $journalEntry->id)
                        ->whereNotIn('id', $existingIds)
                        ->delete();

                    foreach ($accounts as $item) {
                        $ji = !empty($item['id'])
                            ? JournalItem::find($item['id'])
                            : new JournalItem(['journal' => $journalEntry->id]);

                        $ji->fill([
                            'account'     => $item['account'],
                            'description' => $item['description'] ?? null,
                            'debit'       => $item['debit']  ?? 0,
                            'credit'      => $item['credit'] ?? 0,
                        ])->save();

                        Log::info("$sig item saved", ['item_id' => $ji->id]);
                        $this->updateBankBalances($ji);
                    }
                });
                $this->logExecutionTime($t, $sig, 'transactionCommit');

                return redirect()->route(self::INDEX_ROUTE)
                    ->with('success', __('Journal entry successfully updated.'));
            } catch (\Throwable $e) {
                Log::error("$sig error", ['err' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, "$class::$action");
            }
        });
    }

    public function destroy(JournalEntry $journalEntry): RedirectResponse
    {
        $action = __FUNCTION__;
        $class  = static::class;
        $sig    = "$class::$action";

        return $this->measureProfile($action, function () use ($journalEntry, $class, $action, $sig) {
            Log::info("$sig start", ['id' => $journalEntry->id]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if ($c = self::guard(request(), 'delete journal entry', self::INDEX_ROUTE)) {
                Log::warning("$sig denied", ['user' => Auth::id()]);
                return $c;
            }
            if ($journalEntry[DatabaseConstants::TABLE_CREATOR] !== $user?->creatorId()) {
                Log::warning("$sig forbidden", ['owner' => $journalEntry[DatabaseConstants::TABLE_CREATOR]]);
                return defaultPermissionDenial(request(), new AuthorizationException(), "$class::$action");
            }

            try {
                $t = microtime(true);
                DB::transaction(function () use ($journalEntry, $sig) {
                    JournalItem::where('journal', $journalEntry->id)->delete();
                    $journalEntry->delete();
                    Log::info("$sig deleted", ['id' => $journalEntry->id]);
                });
                $this->logExecutionTime($t, $sig, 'transactionCommit');

                return redirect()->route(self::INDEX_ROUTE)
                    ->with('success', __('Journal entry successfully deleted.'));
            } catch (\Throwable $e) {
                Log::error("$sig error", ['err' => $e->getMessage()]);
                return defaultUndefinedException(request(), $e, "$class::$action");
            }
        });
    }

    public const ACC_DST = 'accountDestroy';
    public function accountDestroy(Request $request): RedirectResponse|null
    {
        $action = __FUNCTION__;
        $class  = static::class;
        $sig    = "$class::$action";

        return $this->measureProfile($action, function () use ($request, $class, $action, $sig) {
            Log::info("$sig start", ['input' => $request->all(), 'user_id' => Auth::id()]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, 'delete journal entry', self::INDEX_ROUTE)) !== true) {
                Log::warning("$sig permission denied", ['user_id' => $user?->id]);
                return $c;
            }

            try {
                $t = microtime(true);
                DB::transaction(fn() => JournalItem::where('id', $request->input('id'))->delete());
                $this->logExecutionTime($t, $sig, 'transactionCommit');

                Log::info("$sig deleted journal item", ['item_id' => $request->input('id')]);
                return redirect()->back()
                    ->with('success', __('Journal entry account successfully deleted.'));
            } catch (\Throwable $e) {
                Log::error("$sig failed", ['error' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, "$class::$action", route(self::INDEX_ROUTE));
            }
        });
    }

    public const JRN_DST = 'journalDestroy';
    public function journalDestroy(Request $request, int|string $itemId): RedirectResponse|null
    {
        $action = __FUNCTION__;
        $class  = static::class;
        $sig    = "$class::$action";

        return $this->measureProfile($action, function () use ($request, $itemId, $class, $action, $sig) {
            Log::info("$sig start", ['item_id' => $itemId, 'user_id' => Auth::id()]);
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            $user = $userOrRedirect;
            if (($c = self::guard($request, 'delete journal entry', self::INDEX_ROUTE)) !== true) {
                Log::warning("$sig permission denied", ['user_id' => $user?->id]);
                return $c;
            }

            try {
                $t = microtime(true);
                DB::transaction(function () use ($itemId) {
                    $journalItem = JournalItem::findOrFail($itemId);
                    $journalItem->delete();
                });
                $this->logExecutionTime($t, $sig, 'transactionCommit');

                Log::info("$sig deleted journal item", ['item_id' => $itemId]);
                return redirect()->back()
                    ->with('success', __('Journal account successfully deleted.'));
            } catch (\Illuminate\Database\Eloquent\ModelNotFoundException $e) {
                Log::warning("$sig not found", ['item_id' => $itemId]);
                return redirect()->back()
                    ->with('error', __('Journal account not found.'));
            } catch (\Throwable $e) {
                Log::error("$sig failed", ['error' => $e->getMessage()]);
                return defaultUndefinedException($request, $e, "$class::$action", route(self::INDEX_ROUTE));
            }
        });
    }

    private function updateBankBalances(JournalItem $item): void
    {
        $banks = BankAccount::where('chart_account_id', $item->account)->get();
        foreach ($banks as $bank) {
            $old = $bank->opening_balance;
            $new = $old
                - ($item->debit  ?? 0)
                + ($item->credit ?? 0);
            $bank->opening_balance = $new;
            $bank->save();
            Log::info('bank balance updated', [
                'bank_id'      => $bank->id,
                'new_balance'  => $new
            ]);
        }
    }

    private function journalNumber(): int|string|RedirectResponse
    {
        if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
        $user = $userOrRedirect;

        $latest = JournalEntry::where(DatabaseConstants::TABLE_CREATOR, $user?->creatorId())
            ->latest()
            ->first();

        return $latest
            ? (is_numeric($latest->journal_id) ? $latest->journal_id + 1 : $latest->journal_id)
            : 0;
    }
}

File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/shapes/journal_entry_controller.py
import logging
import inspect
from django.contrib import messages
from django.core.exceptions import PermissionDenied
from django.db import transaction
from django.http import HttpRequest, HttpResponse
from django.shortcuts import render, redirect, get_object_or_404
from .._helpers.http import get_redirect_url
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
from .._traits.controller import Controller
from ....Models.utils.utility import Utility
from ....Models.companies.bank_account import BankAccount
from ....Models.charts.chart_of_account import ChartOfAccount
from ....Models.shapes.journal_entry import JournalEntry
from ....Models.planning.journal_item import JournalItem

logger = logging.getLogger(__name__)


class JournalEntryController(Controller):

    @classmethod
    def _set_entry(cls, entry: JournalEntry, data: dict, user, journal_id=None) -> JournalEntry:
        if journal_id is not None:
            entry.journal_id = journal_id
        for field in ('date', 'reference', 'description'):
            setattr(entry, field, data.get(field))
        entry.created_by = user.creator_id()
        return entry

    @classmethod
    def _process_item(cls, item_data: dict, entry: JournalEntry) -> JournalItem:
        ji = JournalItem.objects.filter(id=item_data.get('id')).first() or JournalItem()
        ji.journal = entry.id
        for field in ('account', 'description', 'debit', 'credit'):
            setattr(ji, field, item_data.get(field, getattr(ji, field, 0)))
        ji.save()
        for ba in BankAccount.objects.filter(chart_account_id=ji.account):
            balance = ba.opening_balance - ji.debit if ji.debit else ba.opening_balance + ji.credit
            ba.opening_balance = balance
            ba.save()
        return ji

    @classmethod
    def index(cls, request: HttpRequest) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            ctrl = cls(); ctrl.request = request; ctrl.authorize('manage journal entry')
            entries = JournalEntry.objects.filter(created_by=request.user.creator_id()).all()
            return render(request, 'journalEntry/index.html', {'journal_entries': entries})
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=f'{CN}::{MN}', logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def create(cls, request: HttpRequest) -> HttpResponse:
        from django.db.models import Value
        from django.db.models.functions import Concat
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            ctrl = cls(); ctrl.request = request; ctrl.authorize('create journal entry')
            accounts = (
                ChartOfAccount.objects.filter(created_by=request.user.creator_id())
                .annotate(code_name=Concat('code', Value(' - '), 'name'))
                .values_list('code_name', 'id')
            )
            journal_id = cls.journal_number(request)
            return render(request, 'journalEntry/create.html', {'accounts': accounts, 'journal_id': journal_id})
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=f'{CN}::{MN}', logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def store(cls, request: HttpRequest) -> HttpResponse:
        import json
        from json import JSONDecodeError
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            ctrl = cls(); ctrl.request = request; ctrl.authorize('create journal entry')
            data = request.POST
            raw = data.get('accounts')
            try:
                accounts = json.loads(raw) if raw else []
            except JSONDecodeError as e:
                return default_permission_denial(request, err=e, ref=f'{CN}::{MN}', logger=logger)
            total_debit = sum(item.get('debit', 0) for item in accounts)
            total_credit = sum(item.get('credit', 0) for item in accounts)
            if total_debit != total_credit:
                messages.error(request, 'Debit and Credit must be Equal.')
                return redirect(get_redirect_url(request))
            with transaction.atomic():
                journal = cls._set_entry(JournalEntry(), data, request.user, cls.journal_number(request))
                journal.save()
                for item in accounts:
                    cls._process_item(item, journal)
            messages.success(request, 'Journal entry successfully created.')
            return redirect('journal_entry_index')
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=f'{CN}::{MN}', logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def show(cls, request: HttpRequest, journal_entry_id: int) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            ctrl = cls(); ctrl.request = request; ctrl.authorize('show journal entry')
            entry = get_object_or_404(JournalEntry, pk=journal_entry_id)
            entry.created_by == request.user.creator_id() or (_ for _ in ()).throw(PermissionDenied())
            accounts = JournalItem.objects.filter(journal=entry.id).all()
            settings = Utility.settings()
            return render(request, 'journalEntry/view.html', {
                'journal_entry': entry,
                'accounts': accounts,
                'settings': settings
            })
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=f'{CN}::{MN}', logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def edit(cls, request: HttpRequest, journal_entry_id: int) -> HttpResponse:
        from django.db.models import Value
        from django.db.models.functions import Concat
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            ctrl = cls(); ctrl.request = request; ctrl.authorize('edit journal entry')
            entry = get_object_or_404(JournalEntry, pk=journal_entry_id)
            accounts = (
                ChartOfAccount.objects.filter(created_by=request.user.creator_id())
                .annotate(code_name=Concat('code', Value(' - '), 'name'))
                .values_list('id', 'id')
            )
            return render(request, 'journalEntry/edit.html', {
                'journal_entry': entry,
                'accounts': accounts
            })
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=f'{CN}::{MN}', logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def update(cls, request: HttpRequest, journal_entry_id: int) -> HttpResponse:
        import json
        from json import JSONDecodeError
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            ctrl = cls(); ctrl.request = request; ctrl.authorize('edit journal entry')
            entry = get_object_or_404(JournalEntry, pk=journal_entry_id)
            data = request.POST
            raw = data.get('accounts')
            try:
                accounts = json.loads(raw) if raw else []
            except JSONDecodeError as e:
                return default_permission_denial(request, err=e, ref=f'{CN}::{MN}', logger=logger)
            total_debit = sum(item.get('debit', 0) for item in accounts)
            total_credit = sum(item.get('credit', 0) for item in accounts)
            if total_debit != total_credit:
                messages.error(request, 'Debit and Credit must be Equal.')
                return redirect(get_redirect_url(request))
            with transaction.atomic():
                cls._set_entry(entry, data, request.user).save()
                for item in accounts:
                    cls._process_item(item, entry)
            messages.success(request, 'Journal entry successfully updated.')
            return redirect('journal_entry_index')
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=f'{CN}::{MN}', logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def destroy(cls, request: HttpRequest, journal_entry_id: int) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            ctrl = cls(); ctrl.request = request; ctrl.authorize('delete journal entry')
            entry = get_object_or_404(JournalEntry, pk=journal_entry_id)
            with transaction.atomic():
                JournalItem.objects.filter(journal=entry.id).delete()
                entry.delete()
            messages.success(request, 'Journal entry successfully deleted.')
            return redirect('journal_entry_index')
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=f'{CN}::{MN}', logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def journal_number(cls, request: HttpRequest) -> int:
        latest = JournalEntry.objects.filter(created_by=request.user.creator_id()).order_by('-id').first()
        return 1 if not latest else latest.journal_id + 1

    @classmethod
    def account_destroy(cls, request: HttpRequest) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            ctrl = cls(); ctrl.request = request; ctrl.authorize('delete journal entry')
            JournalItem.objects.filter(id=request.POST.get('id')).delete()
            messages.success(request, 'Journal entry account successfully deleted.')
            return redirect(get_redirect_url(request))
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=f'{CN}::{MN}', logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @classmethod
    def journal_destroy(cls, request: HttpRequest, item_id: int) -> HttpResponse:
        CN = cls.__name__; MN = inspect.currentframe().f_code.co_name
        try:
            ctrl = cls(); ctrl.request = request; ctrl.authorize('delete journal entry')
            JournalItem.objects.filter(id=item_id).delete()
            messages.success(request, 'Journal account successfully deleted.')
            return redirect(get_redirect_url(request))
        except PermissionDenied as e:
            return default_permission_denial(request, err=e, ref=f'{CN}::{MN}', logger=logger)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)
